src/coprocessor/dag/expr/builtin_op.rs

```diff
@@ -0,0 +1,349 @@
+        Ok(Some(input.map_or(0, |i| (i != 0f64) as i64)))
+    }
+
+    pub fn decimal_is_true(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let input = try!(self.children[0].eval_decimal(ctx, row));
+        Ok(Some(
+            input.map_or(0, |dec| !dec.into_owned().is_zero() as i64),
+        ))
+    }
+
+    pub fn int_is_false(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let input = try!(self.children[0].eval_int(ctx, row));
+        Ok(Some(input.map_or(0, |i| (i == 0) as i64)))
+    }
+
+    pub fn unary_not(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let arg = try_opt!(self.children[0].eval_int(ctx, row));
+        Ok(Some((arg == 0) as i64))
+    }
+
+    pub fn unary_minus_int(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        unimplemented!()
+    }
+
+    pub fn unary_minus_decimal(
+        &self,
+        ctx: &StatementContext,
+        row: &[Datum],
+    ) -> Result<Option<i64>> {
+        unimplemented!()
+    }
+
+    pub fn unary_minus_real(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        unimplemented!()
+    }
+
+    pub fn decimal_is_null(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let arg = try!(self.children[0].eval_decimal(ctx, row));
+        Ok(Some(arg.is_none() as i64))
+    }
+
+    pub fn int_is_null(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let arg = try!(self.children[0].eval_int(ctx, row));
+        Ok(Some(arg.is_none() as i64))
+    }
+
+    pub fn real_is_null(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let arg = try!(self.children[0].eval_real(ctx, row));
+        Ok(Some(arg.is_none() as i64))
+    }
+
+    pub fn string_is_null(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let arg = try!(self.children[0].eval_string(ctx, row));
+        Ok(Some(arg.is_none() as i64))
+    }
+
+    pub fn time_is_null(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let arg = try!(self.children[0].eval_time(ctx, row));
+        Ok(Some(arg.is_none() as i64))
+    }
+
+    pub fn duration_is_null(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let arg = try!(self.children[0].eval_duration(ctx, row));
+        Ok(Some(arg.is_none() as i64))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use tipb::expression::ScalarFuncSig;
+    use coprocessor::codec::Datum;
+    use coprocessor::codec::mysql::Duration;
+    use coprocessor::dag::expr::{Expression, StatementContext};
+    use coprocessor::dag::expr::test::{fncall_expr, str2dec};
+    use coprocessor::select::xeval::evaluator::test::datum_expr;
+
+    #[test]
+    fn test_logic_op() {
+        let tests = vec![
+            (
+                ScalarFuncSig::LogicalAnd,
```

Comment:
You can reduce cases by always checking result is the same after switching lhs and rhs.